File: LeetCode/0053. Maximum Subarray/solution.cpp
/**
 * 202 / 202 test cases passed.
 * Status: Accepted
 * Runtime: 16 ms
 */
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        vector<int> dp( nums.size(), 0 );
        dp[0] = nums[0];
        int ans = dp[0];
        for ( int i = 1; i < nums.size(); ++ i) {
            dp[i] = max( dp[i - 1] + nums[i], nums[i] );
            ans = max(ans, dp[i]);
        }
        return ans;
    }
};

File: LeetCode/0053. Maximum Subarray/solution.go
/**
 * 203 / 203 test cases passed.
 * Runtime: 4 ms
 * Memory Usage: 3.4 MB
 */
func maxSubArray(nums []int) int {
	dp := make([]int, len(nums)+1)
	ans := nums[0]
	for i, num := range nums {
		dp[i+1] = max(dp[i]+num, num)
		ans = max(ans, dp[i+1])
	}
	return ans
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
